pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,552 @@
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem
+// function in order to read and process a message on the workqueue.
+func (c *MCRouteController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MCRouteController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if k, ok := obj.(key); !ok {
+		c.queue.Forget(obj)
+		klog.InfoS("Expected type 'key' in work queue but got object", "object", obj)
+		return true
+	} else if err := c.syncMCFlows(k); err == nil {
+		c.queue.Forget(k)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCRouteController) syncMCFlows(k key) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing Gateway Node flows for Multi-cluster", "type", k.kind,
+			"name", k.name, "time", time.Since(startTime))
+	}()
+	switch k.kind {
+	case gatewayKind:
+		klog.V(2).InfoS("Processing for Gateway", "gateway", k.name)
+		return c.handleGatewayEvent(k)
+	case ciImportKind:
+		klog.V(2).InfoS("Processing for ClusterInfoImport", "clusterinfoimport", k.name)
+		return c.handleCIImportEvent(k)
+	}
+	return nil
+}
+
+func (c *MCRouteController) handleGatewayEvent(k key) error {
+	var gwIsActive, iAmGW, noActiveGateway bool
+	var activeGW *mcv1alpha1.Gateway
+	gws, err := c.getGateways()
+	if err != nil {
+		return err
+	}
+	if gws != nil {
+		activeGW = c.getActiveGateway(gws)
+		klog.V(2).InfoS("The active Gateway is found", "gateway", klog.KObj(activeGW))
+		iAmGW = c.iAmGateway(gws)
+		gwIsActive = activeGW.Name == k.name
+		c.iAmActiveGW = activeGW.Name == c.nodeConfig.Name
+	} else {
+		noActiveGateway = true
+		c.iAmActiveGW = false
+		iAmGW = c.nodeConfig.Name == k.name
+	}
+
+	_, err = c.gwLister.Gateways(c.namespace).Get(k.name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			if err := c.cleanupFlows(k); err != nil {
+				return err
+			}
+			// When there is no active Gateway after a Gateway deletion, just return.
+			// Otherwise, we need reinstall multi-cluster flows for the latest active Gateway
+			if noActiveGateway {
+				return nil
+			}
+			klog.V(2).InfoS("Installing multi-cluster flows for the active Gateway", "gateway", klog.KObj(activeGW))
+			if !c.iAmActiveGW {
+				return c.installFlowsOnRegularNode(activeGW)
+			}
+			return c.installFlowsOnGateway(activeGW)
+		}
+		return err
+	}
+
+	if c.iAmActiveGW && (activeGW.Name != c.installedGWName && c.installedGWName != "") {
```

Comment:
For @tnqn 's proposal, the only thing is we need to handle rollback/cleanup of partially installed flows at errors, since we change installedActiveGWName based on error or not. Or should we add one more flag about all flows are installed or not to be safer?

Reply:
For @tnqn 's proposal, the only thing is we need to handle partial success of flow installation. We should at least try rolling back, but to be safer (assuming rollback can fail too), we might need to save a flag about flows are all installed or not?

Reply:
Maybe we could use `c.installedActiveGWName` to represent the flows for gateway/normalNode has been installed successfully and items of `c.installedClusterInfoImports` to represent the flows for a remote cluster has been installed successfully, then the sync function should be idempotent and could simply return error and retry later once OVS call fails.

Reply:
If the flows added for clusterInfoImports are not relevant to GWs, and do not differ on GW and regular Nodes we can do that. Are these true? I need to check.

Reply:
The flows are different on GW/regular Nodes and are relevant to GWs. I guess we might need a flag per CIImport to indicate error or not, and the new round should first try removing flows if error.

Reply:
For flows installed for CIImport, there should be no problem because if gateway changes, all flows for installed CIImports will be removed anyway, it should use `installedCIImports` when cleaning up, each item in the list/map represent it was installed successfully.
For flows installed for Gateway/NormalNode, it should be the same?

Reply:
I got the problem you concerned, I think we could remove gateway name from the key of the flows installed for CIImports?

Reply:
Even if gateway name is kept in CIImport flows, it may still work, as long as `installedActiveGWName` is in sync with the flows. The sync function should always do cleanup first before installing new flows, so it could use previous gateway name to remove CIImport flows.

Reply:
yes, not matter it's Gateway or CIImport events, it will use `installedCIImports` to clean up previous installed flows, I am checking how to simplify this part, thanks for all valuable suggestions.